backend/hitas/services/reports.py

```diff
@@ -0,0 +1,227 @@
+import datetime
+import string
+from decimal import Decimal
+from functools import cache
+from statistics import mean
+from typing import Any, Callable, Iterable, NamedTuple, TypeVar
+
+from openpyxl.styles import Alignment, Border, Side
+from openpyxl.workbook import Workbook
+from openpyxl.worksheet.worksheet import Worksheet
+from rest_framework.exceptions import ValidationError
+from rest_framework.settings import api_settings
+
+from hitas.models import ApartmentSale
+from hitas.utils import format_sheet, resize_columns
+
+
+class SalesReportColumns(NamedTuple):
+    cost_area: int | str
+    postal_code: str
+    apartment_address: str
+    notification_date: datetime.date | str
+    purchase_date: datetime.date | str
+    purchase_price_per_square_meter: Decimal | str
+    total_price_per_square_meter: Decimal | str
+    exclude_from_statistics: bool | str
+
+
+T = TypeVar("T")
+
+
+class SalesReportSummaryDefinition(NamedTuple):
+    func: Callable[[Iterable[T]], T]
+    title: str = ""
+    subtitle: str = ""
+
+
+def build_sales_report_excel(sales: list[ApartmentSale]) -> Workbook:
+    workbook = Workbook()
+    worksheet: Worksheet = workbook.active
+
+    column_headers = SalesReportColumns(
+        cost_area="Kalleusalue",
+        postal_code="Postinumero",
+        apartment_address="Osoite",
+        notification_date="Ilmoituspäivä",
+        purchase_date="Kauppapäivä",
+        purchase_price_per_square_meter="Kaupan neliöhinta",
+        total_price_per_square_meter="Velaton neliöhinta",
+        exclude_from_statistics="Ei tilastoihin",
+    )
+    worksheet.append(column_headers)
+
+    for sale in sales:
+        if not sale.apartment.surface_area:
+            raise ValidationError(
+                detail={
+                    api_settings.NON_FIELD_ERRORS_KEY: (
+                        f"Surface are zero or missing for apartment {sale.apartment.address!r}. "
+                        f"Cannot calculate price per square meter."
+                    )
+                },
+            )
+
+        worksheet.append(
+            SalesReportColumns(
+                cost_area=sale.apartment.postal_code.cost_area,
+                postal_code=sale.apartment.postal_code.value,
+                apartment_address=sale.apartment.address,
+                notification_date=sale.notification_date,
+                purchase_date=sale.purchase_date,
+                purchase_price_per_square_meter=sale.purchase_price / sale.apartment.surface_area,
+                total_price_per_square_meter=sale.total_price / sale.apartment.surface_area,
+                exclude_from_statistics=sale.exclude_from_statistics,
```

Comment:
I would change the "Ei Tilastoihin" value from "TRUE/FALSE" to something more descriptive. e.g. Only showing TRUE values

Reply:
I can put X / " " there also.

Reply:
Actually, we changed this already so that excluded sales are not in this report, based on the customer's feedback, so this column should be removed.

Reply:
Field removed